src/coreclr/gc/unix/gcenv.unix.cpp

```diff
@@ -868,7 +868,7 @@ bool ReadMemoryValueFromFile(const char* filename, uint64_t* val)
     return result;
 }
 
-#define UPDATE_CACHE_SIZE_AND_LEVEL(NEW_CACHE_SIZE, NEW_CACHE_LEVEL) if (NEW_CACHE_SIZE > ((long)cacheSize)) { cacheSize = NEW_CACHE_SIZE; cacheLevel = NEW_CACHE_LEVEL; }
+#define UPDATE_CACHE_SIZE_AND_LEVEL(CURRENT_CACHE_SIZE, NEW_CACHE_SIZE, NEW_CACHE_LEVEL) if (NEW_CACHE_SIZE > ((long)CURRENT_CACHE_SIZE)) { CURRENT_CACHE_SIZE = NEW_CACHE_SIZE; cacheLevel = NEW_CACHE_LEVEL; }
```

Comment:
The `cacheLevel` still stayed as an ambient variable modified by the macro. I would actually recommend replacing the macro by a function instead and pass in the new values and pointers to the current values instead?

Reply:
I agree that it would be best to get rid of the macro. I would convert it to a for-loop. Also, I may be a good idea to run through the array backwards to avoid unnecessary queries for caches of lower levels.

```
#if HAVE_SYSCONF

const int cacheLevelNames[] = {
    _SC_LEVEL1_DCACHE_SIZE,
    _SC_LEVEL2_DCACHE_SIZE,
    _SC_LEVEL3_DCACHE_SIZE,
    _SC_LEVEL4_DCACHE_SIZE,
}

for (int i = ARRAY_SIZE(cacheLevelNames) - 1; i >= 0; i--)
{
    long size = sysconf(cacheLevelNames[i]);
    if (size > 0)
    {
        cacheSize = (size_t)size;
        cacheLevel = i + 1;
        break;
    }
}

#endif // HAVE_SYSCONF
```